File: c/merge_two_sorted_lists.c
// Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include <stdio.h>

#include "lib/list.h"

typedef struct ListNode* Position;
typedef struct ListNode* List;

struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
	List L;
	Position P;

	if (l1 == NULL)
		return l2;

	if (l2 == NULL)
		return l1;

	if (l1 -> val > l2 -> val)
	{
		L = l2;
		l2 = l2 -> next;
	}
	else
	{
		L = l1;
		l1 = l1 -> next;
	}
	
	P = L;

	while (l2 != NULL && l1 != NULL)
	{
		if (l1 -> val < l2 -> val)
		{
			P -> next = l1;
			l1 = l1 -> next;
		}
		else
		{
			P -> next = l2;
			l2 = l2 -> next;
		}

		P = P -> next;
	}

	if (l1 != NULL)
		P -> next = l1;

	if (l2 != NULL)
		P -> next = l2;

	return L;
}

int main(int argc, char const *argv[])
{
	List L1, L2, L;

	L1 = CreateEvenList(10);
	L2 = CreateOddList(10);
	PrintList(L1);
	PrintList(L2);

	L = mergeTwoLists(L1, L2);
	PrintList(L);
	return 0;
}

File: javascript/merge_two_sorted_lists.js
// Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var listModule = require('./src/list.js');

var mergeTwoLists = function(headA, headB) {
	if (headA === null || headA === undefined)
		return headB;

	if (headB === null || headB === undefined)
		return headA;

	var head;

	if (headA.val < headB.val)
	{
		head = headA;
		headA = headA.next;
	}
	else
	{
		head = headB;
		headB = headB.next;
	}

	node = head;
	while (headA !== null && headB !== null)
	{
		if (headA.val < headB.val)
		{
			node.next = headA;
			headA = headA.next;
		}
		else
		{
			node.next = headB;
			headB = headB.next;
		}

		node = node.next;
	}

	if (headA !== null)
		node.next = headA;

	if (headB !== null)
		node.next = headB;

	return head;
};

var head1 = listModule.createList(10);
var head2 = listModule.createList(10);

listModule.printList(head1);
listModule.printList(head2);

var head = mergeTwoLists(head1, head2);
listModule.printList(head);


